### PR TITLE
Add quick start gNMI/RESTCONF example app - 14.2.x

### PR DESCRIPTION
**Pull request by PeterSuna**
- Add gNMI/RESTCONF app with quick start example which could be started and use.
- Describe how to use and test example.
- Add postman collection with requests example.
- Add test which verify that test scenario is working.

### CODE REVIEW

File: lighty-examples/lighty-gnmi-community-restconf-app/README.md [hunk 1]
```diff
@@ -0,0 +1,81 @@
+# lighty.io gNMI/RESTCONF application
+This application provides gNMI south-bound utilized with RESTCONF interface to manage gNMI devices on the network.
+Application works as standalone SDN controller. It is capable to connect to gNMI devices and expose connected devices
+over RESTCONF north-bound APIs. In this application will be started gNMI simulator as gNMI target and all operation will
```
Comment (martinbugan-pt) [inline lighty-examples/lighty-gnmi-community-restconf-app/README.md:4]: In this application gNMI simulator will be started as gNMI target and all operations will..

File: lighty-examples/lighty-gnmi-community-restconf-app/README.md [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+be performed on this device.
+
+Application lighty.io gNMI/RESTCONF is pre-prepared with [Openconfig YANG models](src/main/assembly/resources/yangs).
+These models are used by both gNMI application and gNMI device simulator. Device have already prepared state/config data
```
Comment (martinbugan-pt) [inline lighty-examples/lighty-gnmi-community-restconf-app/README.md:8]: Device has already..

File: lighty-examples/lighty-gnmi-community-restconf-app/README.md [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+configured by [this](src/main/assembly/resources/simulator/initialJsonData.json) json file.
+For communication with gNMI device is required to use TLS communication with certificates and authorized
```
Comment (martinbugan-pt) [inline lighty-examples/lighty-gnmi-community-restconf-app/README.md:10]: with gNMI device there is required to use...

File: lighty-examples/lighty-gnmi-community-restconf-app/README.md [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
+
+2. Unpack the application ZIP distribution created in the _lighty-gnmi-community-restconf-app/target_ called
+   `lighty-gnmi-community-restconf-app-<version>-bin.zip`
+
+3. Start the application by:
+   - Running it's _.jar_ file `java -jar lighty-gnmi-community-restconf-app-<version>-SNAPSHOT.jar`
+   - Or with provided script `./start-controller.sh`
+
+## Example of using lighty.io gNMI/RESTCONF application
+This example show how to connect gNMI device and perform basic CRUD operation on the device. All RESTCONF request
```
Comment (martinbugan-pt) [inline lighty-examples/lighty-gnmi-community-restconf-app/README.md:38]: example shows..  ..basic CRUD operations.. .. RESTCONF requests

File: lighty-examples/lighty-gnmi-community-restconf-app/README.md [hunk 5]
```diff
@@ -0,0 +1,81 @@
...
+used in this example are provided in [postman-collection](lighty.io gNMI-RESTCONF application.postman_collection.json).
+
+ - ### Add client certificates to lighty.io gNMI keystore
+Used certificates can be found [here](src/main/assembly/resources/certificates). To keystore is added only client
```
Comment (martinbugan-pt) [inline lighty-examples/lighty-gnmi-community-restconf-app/README.md:42]: only client certificates are added to keystore

File: lighty-examples/lighty-gnmi-community-restconf-app/README.md [hunk 6]
```diff
@@ -0,0 +1,81 @@
...
+certificates. Adding required certificates for gNMI device to lighty.io gNMI application is performed by
+postman request `'Add Keystore'`.
+
+ - ### Connect gNMI device
+Device connection is performed by request `'Connet device'`. In the body of this request is contained identifier
```
Comment (martinbugan-pt) [inline lighty-examples/lighty-gnmi-community-restconf-app/README.md:47]: Connect device.. ..of this request, identifier is contained for..

File: lighty-examples/lighty-gnmi-community-restconf-app/README.md [hunk 7]
```diff
@@ -0,0 +1,81 @@
...
+for keystore, device information, extension parameters and basic authorization required by device. When device
+is successfully connected to application, then these logs should be visible:
+```
+ INFO [gnmi_executor-1] (GnmiMountPointRegistrator.java:52) - Mount point for node gnmi-simulator created: {closed=false, instance=org.opendaylight.mdsal.dom.spi.SimpleDOMMountPoint@60112bc5}
+ INFO [gnmi_executor-0] (GnmiNodeListener.java:105) - Connection with node Uri{_value=gnmi-simulator} established successfully
+```
+
+Device state can be also checked by request `'Get gnmi-simulator node'`. Inside this request can be found information about
```
Comment (martinbugan-pt) [inline lighty-examples/lighty-gnmi-community-restconf-app/README.md:55]: information about created device can be found inside this request

File: lighty-examples/lighty-gnmi-community-restconf-app/README.md [hunk 8]
```diff
@@ -0,0 +1,81 @@
...
+created device. In section `gnmi-topology:node-state` can be found current state of device or information about error
```
Comment (martinbugan-pt) [inline lighty-examples/lighty-gnmi-community-restconf-app/README.md:56]: In section gnmi-topology:node-state, current state of device or information about occured errors can be found.

File: lighty-examples/lighty-gnmi-community-restconf-app/README.md [hunk 9]
```diff
@@ -0,0 +1,81 @@
...
+ - ### Update data on gNMI device with PATCH request
+For updating data send request `'Update config data'`. This request will append new type to config authentication-method.
+To validate request send GET request `'Get Authentication from CONFIG'`.
+
+ - ### Delete data on gNMI device
+For deleting `config` container send request `'Delete authentication config'`. To validate request send GET
+request `'Get Authentication from CONFIG'`.
+
+ - ### Remove gNMI device from lighty.io gNMI/RESTCONF application
+When is required to restart connection or just remove device send request `'Remove device'`.
```
Comment (martinbugan-pt) [inline lighty-examples/lighty-gnmi-community-restconf-app/README.md:80]: When restart of connection or removal of device is required, just send request

File: lighty-examples/lighty-gnmi-community-restconf-app/src/test/java/io/lighty/examples/controllers/gnmi/GnmiRestconfExampleTest.java [hunk 10]
```diff
@@ -0,0 +1,312 @@
...
+    private static final String GET_AUTHENTICATION_CONFIG = GET_AUTHENTICATION + "/config" ;
+    private static final String GET_CONFIG_AUTHENTICATION_CONFIG = GET_AUTHENTICATION_CONFIG + "?content=config" ;
+    private static final String GET_CONFIG_AUTHENTICATION = GET_AUTHENTICATION + "?content=config";
+
+    private static ExecutorService httpClientExecutor;
+    private static HttpClient httpClient;
+
+    @BeforeAll
+    public static void startUp() throws IOException {
+        httpClientExecutor = Executors.newSingleThreadExecutor();
```
Comment (martinbugan-pt) [inline lighty-examples/lighty-gnmi-community-restconf-app/src/test/java/io/lighty/examples/controllers/gnmi/GnmiRestconfExampleTest.java:69]: these startup, teardown methods and also tests should not be throwing exceptions, handle them appropriately, also in tests

File: lighty-examples/lighty-gnmi-community-restconf-app/src/test/java/io/lighty/examples/controllers/gnmi/GnmiRestconfExampleTest.java [hunk 11]
```diff
@@ -0,0 +1,312 @@
...
+        sendDeleteRequest(MOUNTPOINT_PATH);
+        httpClientExecutor.shutdownNow();
+    }
+
+    @Test
+    public void readmeExampleTest() throws IOException, InterruptedException {
+        // Register keystore.
+        final String certificatesRequestBody = getCertificatesRequestBody(getResources(CA_CERTIFICATE_PATH),
+                getResources(CLIENT_KEY_PATH), getResources(CLIENT_CERTIFICATE_PATH));
+        final HttpResponse<String> response = sendPostRequest(ADD_CERTIFICATE_PATH, certificatesRequestBody);
```
Comment (martinbugan-pt) [inline lighty-examples/lighty-gnmi-community-restconf-app/src/test/java/io/lighty/examples/controllers/gnmi/GnmiRestconfExampleTest.java:86]: use more descriptive names for variables

File: lighty-examples/lighty-gnmi-community-restconf-app/src/test/java/io/lighty/examples/controllers/gnmi/GnmiRestconfExampleTest.java [hunk 12]
```diff
@@ -0,0 +1,312 @@
...
+        assertEquals(response.statusCode(), HttpURLConnection.HTTP_NO_CONTENT);
+
+        // Check if device was not created before.
+        HttpResponse<String> notCreatedDevice = sendGetRequest(MOUNTPOINT_STATUS_PATH);
```
Comment (martinbugan-pt) [inline lighty-examples/lighty-gnmi-community-restconf-app/src/test/java/io/lighty/examples/controllers/gnmi/GnmiRestconfExampleTest.java:90]: for responses maybe add *Response to be more readable

File: lighty-examples/lighty-gnmi-community-restconf-app/src/test/java/io/lighty/examples/controllers/gnmi/GnmiRestconfExampleTest.java [hunk 13]
```diff
@@ -0,0 +1,312 @@
...
+        if (notCreatedDevice.statusCode() == HttpURLConnection.HTTP_OK) {
+            // Delete device if was created before test.
+            sendDeleteRequest(MOUNTPOINT_PATH);
+        }
+
+        // Register device.
+        final String mRegBody = getMountpointRegistrationBody();
```
Comment (martinbugan-pt) [inline lighty-examples/lighty-gnmi-community-restconf-app/src/test/java/io/lighty/examples/controllers/gnmi/GnmiRestconfExampleTest.java:97]: also here, do not necessarily shorten names of variables, it makes the code a bit less readable in cases like this, use whole mountpointRegistrationBody

**Review by martinbugan-pt — CHANGES_REQUESTED** [review]

File: lighty-examples/lighty-gnmi-community-restconf-app/src/main/java/io/lighty/examples/controllers/gnmi/Main.java [hunk 14]
```diff
@@ -0,0 +1,40 @@
...
+        String currentFolder = System.getProperty(USER_DIR);
+        // Find current location of resources
+        final String gnmiConfig;
+        if (currentFolder.length() > 6 && currentFolder.endsWith(LIGHTY)) {
+            currentFolder += ASSEMBLY_RESOURCES;
+            // Use different gNMI configuration when is example start in IDEA
+            gnmiConfig = ASSEMBLY_RESOURCES + GNMI_CONFIGURATION;
+        } else {
+            gnmiConfig = currentFolder + GNMI_CONFIGURATION;
+        }
```
[the remark this answers is not in the document]
Comment (PeterSuna) [reply lighty-examples/lighty-gnmi-community-restconf-app/src/main/java/io/lighty/examples/controllers/gnmi/Main.java]: done

File: lighty-examples/lighty-gnmi-community-restconf-app/README.md (hunk 2 above)
Comment (PeterSuna) [reply lighty-examples/lighty-gnmi-community-restconf-app/README.md]: done

File: lighty-examples/lighty-gnmi-community-restconf-app/README.md (hunk 3 above)
Comment (PeterSuna) [reply lighty-examples/lighty-gnmi-community-restconf-app/README.md]: done

File: lighty-examples/lighty-gnmi-community-restconf-app/README.md (hunk 4 above)
Comment (PeterSuna) [reply lighty-examples/lighty-gnmi-community-restconf-app/README.md]: done

File: lighty-examples/lighty-gnmi-community-restconf-app/README.md (hunk 5 above)
Comment (PeterSuna) [reply lighty-examples/lighty-gnmi-community-restconf-app/README.md]: done

File: lighty-examples/lighty-gnmi-community-restconf-app/README.md (hunk 6 above)
Comment (PeterSuna) [reply lighty-examples/lighty-gnmi-community-restconf-app/README.md]: done

File: lighty-examples/lighty-gnmi-community-restconf-app/README.md (hunk 7 above)
Comment (PeterSuna) [reply lighty-examples/lighty-gnmi-community-restconf-app/README.md]: done

File: lighty-examples/lighty-gnmi-community-restconf-app/README.md (hunk 8 above)
Comment (PeterSuna) [reply lighty-examples/lighty-gnmi-community-restconf-app/README.md]: done

File: lighty-examples/lighty-gnmi-community-restconf-app/README.md (hunk 9 above)
Comment (PeterSuna) [reply lighty-examples/lighty-gnmi-community-restconf-app/README.md]: done

File: lighty-examples/lighty-gnmi-community-restconf-app/src/test/java/io/lighty/examples/controllers/gnmi/GnmiRestconfExampleTest.java (hunk 11 above)
Comment (PeterSuna) [reply lighty-examples/lighty-gnmi-community-restconf-app/src/test/java/io/lighty/examples/controllers/gnmi/GnmiRestconfExampleTest.java]: done

File: lighty-examples/lighty-gnmi-community-restconf-app/src/test/java/io/lighty/examples/controllers/gnmi/GnmiRestconfExampleTest.java (hunk 12 above)
Comment (PeterSuna) [reply lighty-examples/lighty-gnmi-community-restconf-app/src/test/java/io/lighty/examples/controllers/gnmi/GnmiRestconfExampleTest.java]: done

File: lighty-examples/lighty-gnmi-community-restconf-app/src/test/java/io/lighty/examples/controllers/gnmi/GnmiRestconfExampleTest.java (hunk 13 above)
Comment (PeterSuna) [reply lighty-examples/lighty-gnmi-community-restconf-app/src/test/java/io/lighty/examples/controllers/gnmi/GnmiRestconfExampleTest.java]: done

**Review by martinbugan-pt — APPROVED** [review]

**Review by vainyksi — CHANGES_REQUESTED** [review]

File: lighty-examples/lighty-gnmi-community-restconf-app/README.md [hunk 15]
```diff
@@ -0,0 +1,81 @@
...
+For updating data send request `'Update config data'`. This request will append new type to config authentication-method.
+To validate request send GET request `'Get Authentication from CONFIG'`.
+
+ - ### Delete data on gNMI device
+For deleting `config` container send request `'Delete authentication config'`. To validate request send GET
+request `'Get Authentication from CONFIG'`.
+
+ - ### Remove gNMI device from lighty.io gNMI/RESTCONF application
+When restart of connection or removal of device is required, just send request `'Remove device'`.
+This will remove connected device. For restarting connection it will be required to send request `'Connet device'`
```
[the remark this answers is not in the document]
Comment (PeterSuna) [reply lighty-examples/lighty-gnmi-community-restconf-app/README.md]: done

File: lighty-examples/lighty-gnmi-community-restconf-app/README.md [hunk 16]
```diff
@@ -0,0 +1,81 @@
...
+ INFO [gnmi_executor-0] (GnmiNodeListener.java:105) - Connection with node Uri{_value=gnmi-simulator} established successfully
+```
+
+Device state can be also checked by request `'Get gnmi-simulator node'`. Information about created device can be found
+inside this request. In section `gnmi-topology:node-state`, current state of device or information about occured errors
+can be found. If device is connected, then `node-status` in this response should have value `READY`.
+
+ - ### Get data from gNMI device
+In provided postman-collection few examples of getting data from gNMI device `'Get interfaces'`,`'Get system'` and
+`'Get Authentication'`. In next section for modifying data, we will more focus on system data obtained from
```
[the remark this answers is not in the document]
Comment (PeterSuna) [reply lighty-examples/lighty-gnmi-community-restconf-app/README.md]: done

File: lighty-examples/lighty-gnmi-community-restconf-app/README.md [hunk 17]
```diff
@@ -0,0 +1,81 @@
...
+   - Or with provided script `./start-controller.sh`
+
+## Example of using lighty.io gNMI/RESTCONF application
+This example shows how to connect gNMI device and perform basic CRUD operations on the device. All RESTCONF requests
+used in this example are provided in [postman-collection](lighty.io gNMI-RESTCONF application.postman_collection.json).
+
+ - ### Add client certificates to lighty.io gNMI keystore
+Used certificates can be found [here](src/main/assembly/resources/certificates). Only client certificates are added
+to keystore. Adding required certificates for gNMI device to lighty.io gNMI application is performed by
+postman request `'Add Keystore'`.
```
[the remark this answers is not in the document]
Comment (PeterSuna) [reply lighty-examples/lighty-gnmi-community-restconf-app/README.md:45]: done

File: lighty-examples/lighty-gnmi-community-restconf-app/README.md [hunk 18]
```diff
@@ -0,0 +1,81 @@
...
+## Build and start
+To build and start the lighty.io gNMI/RESTCONF application in your local environment, follow these steps:
+1. Build the application using maven
+   `mvn clean install`
+
+2. Unpack the application ZIP distribution created in the _lighty-gnmi-community-restconf-app/target_ called
+   `lighty-gnmi-community-restconf-app-<version>-bin.zip`
+
+3. Start the application by:
+   - Running it's _.jar_ file `java -jar lighty-gnmi-community-restconf-app-<version>-SNAPSHOT.jar`
```
[the remark this answers is not in the document]
Comment (PeterSuna) [reply lighty-examples/lighty-gnmi-community-restconf-app/README.md]: done

File: lighty-examples/lighty-gnmi-community-restconf-app/src/main/java/io/lighty/examples/controllers/gnmi/Main.java [hunk 19]
```diff
@@ -0,0 +1,73 @@
...
+    private static final String ASSEMBLY_RESOURCES
+            = "/lighty-examples/lighty-gnmi-community-restconf-app/src/main/assembly/resources";
+    private static final String LIGHTY = "lighty";
+    private static final String USER_DIR = "user.dir";
+
+    public static void main(String[] args) throws IOException {
+        final GnmiRestconfAppResources exampleResources = getGnmiRestconfAppResources();
+        GnmiSimulatorApp.main(new String[]{exampleResources.pathToExampleResources});
+        RCgNMIApp.main(new String[]{"-c", exampleResources.gnmiConfigPath});
+    }
```
Comment (vainyksi) [inline lighty-examples/lighty-gnmi-community-restconf-app/src/main/java/io/lighty/examples/controllers/gnmi/Main.java:28]: This start procedure should be refactored similar to https://github.com/PANTHEONtech/lighty/pull/752/commits/b8ba6a640469a95e04b06a0801bdb37a88051f31
The refactor will need the update/refactor of gNMI YANG modules loading

**Review by vainyksi — APPROVED** [review]

**Review by vainyksi — COMMENTED** [review]
This start procedure should be refactored similar to https://github.com/PANTHEONtech/lighty/pull/752/commits/b8ba6a640469a95e04b06a0801bdb37a88051f31
The refactor will need the update/refactor of gNMI YANG modules loading

**Review by vainyksi — APPROVED** [review]

**Pull request merged**